### PR TITLE
Add filter plugin

### PR DESCRIPTION
**Pull request by okkez**

### CODE REVIEW

**Comment by y-ken** [comment]
Hi @okkez 

Thank you for your nice work!
I'll check the code.

**Comment by okkez** [comment]
TODO

* [ ] Extract common functionalities

See also #17

File: lib/fluent/plugin/out_geoip.rb [hunk 1]
```diff
@@ -102,82 +56,7 @@ def shutdown
...
-    return geodata
-  end
-
-  def create_placeholder(geodata)
-    placeholder = {}
-    @placeholder_keys.each do |placeholder_key|
-      position = placeholder_key.match(REGEXP_PLACEHOLDER_SINGLE)
-      next if position.nil? or geodata[position[:record_key]].nil?
-      placeholder.store(placeholder_key, geodata[position[:record_key]][position[:geoip_key].to_sym])
+      Fluent::Engine.emit(tag, time, @supplement.add_geoip_field(record))
```
Comment (cosmo0920) [inline lib/fluent/plugin/out_geoip.rb]: `Fluent::Engine.emit` can replace `router.emit` in fluentd 0.10.58 or later.
see: https://github.com/fluent/fluentd/blob/v0.10/ChangeLog#L40

Comment (repeatedly) [inline lib/fluent/plugin/out_geoip.rb]: For performance improvement, emit_stream with MultiEventStream is better.

**Comment by okkez** [comment]
@repeatedly I think performance improvement is not purpose of this issue.

**Comment by y-ken** [comment]
Looks nice!

**Comment by okkez** [comment]
Sorry for my mistake.

@y-ken Please merge this PR.

**Comment by repeatedly** [comment]
@y-ken How about this?

**Comment by repeatedly** [comment]
ping?
other user relesaed another filter plugin.
I think, this is very confused > https://github.com/bungoume/fluent-plugin-geoip-filter

**Comment by okkez** [comment]
@y-ken Do you have any concern?

**Comment by okkez** [comment]
ping? @y-ken 
Do you have any concern?

**Comment by y-ken** [comment]
Hi @okkez 
I'm very sorry for taking a long time to check. It looks good to me.
But I felt some opinion for this.

* Please add `define_method("router")` to support Fluentd v0.10.57 or earlier
cf. https://github.com/y-ken/fluent-plugin-twitter/blob/89fd2f0e2611b9c26da2930ce1a47cc6f128a86a/lib/fluent/plugin/in_twitter.rb#L7

* On this change has merged, Would you please give me a PR which implement with `emit_stream ` instead of `emit` to improve performance?
https://github.com/y-ken/fluent-plugin-geoip/pull/24#discussion-diff-36387192

* Do you have any idea for the class name of GeoIPSupplement?
Of course it is very little problem. I feel something , utility class style name or service class is better than suffix of `Supplement`. I cannot decide what is the best for this time.

**Comment by okkez** [comment]
Thank you for merging this PR.

> Please add define_method("router") to support Fluentd v0.10.57 or earlier

OK, I will add it later in this PR.

> On this change has merged, Would you please give me a PR which implement with emit_stream instead of emit to improve performance?

Sure! I will do that after merge this PR.

> Do you have any idea for the class name of GeoIPSupplement?

No, I don't, for now. But I will consider class name.

**Comment by okkez** [comment]
>>  Do you have any idea for the class name of GeoIPSupplement?

> No, I don't, for now. But I will consider class name.

How about just `Fluent::GeoIP`?

**Comment by y-ken** [comment]
`Fluent::GeoIP` looks good for me!
Would you please `git cherry-pick` from #29 to fix travis-ci error?

**Comment by cosmo0920** [comment]
I think that rebasing with current master's HEAD is another solution.

**Comment by y-ken** [comment]
Thank you very much!

> I think that rebasing with current master's HEAD is another solution.
exactly.

**Comment by y-ken** [comment]
I'll merge this PR. Any concern?
ping @repeatedly

**Comment by y-ken** [comment]
Merged! Everyone, thank you so very much.

**Comment by repeatedly** [comment]
One user hits output filter approach issue.
Releasing new version is better for newbies.

https://groups.google.com/forum/#!topic/fluentd/vUgObn0JraY

**Comment by y-ken** [comment]
OK. I'll release it now.

**Comment by y-ken** [comment]
@okkez 
Would you please update README how to use filter plugin?

**Comment by okkez** [comment]
OK. I will write documents in next week.

**Comment by y-ken** [comment]
@okkez Thank you very much!

**Comment by y-ken** [comment]
I have released it now.
https://rubygems.org/gems/fluent-plugin-geoip

After next week, I'd like to release performance improvement patch by @okkez .

**Pull request merged**